Code-4-Community/speak-for-the-trees-frontend
# Improve after-login redirect

## Why

[ClickUp Ticket](https://app.clickup.com/t/2b93c6u)



After a user logs in, they're redirected from the Login page to their original destination (or the Home page if they had none). The way this originally was done meant trying to render the Login page while updating the state and rendering the new page (basically, a mess). 

<img height=400px src=https://user-images.githubusercontent.com/22990100/154177430-06d1137b-70f5-4d26-bb09-b23ec606d36a.png />

This ticket improves how we handle redirect and login to get rid of this error.

## This PR



- Instead of using `useHistory` to push the user to a new location, the Login page simply returns a `Redirect` (and, more importantly, doesn't return the login page at the same time!)
- We have to `getUserData` to show the user's name on the home page and in the nav bar. Originally, we did this in the login page after they logged in. However, it makes a lot more sense to get that user data before the user is redirected to the new page (ex. if the data doesn't load in time, the name will just be missing). This ticket adds a call to get the user data in the `login` thunk so that we always get the user data at the same time as log in (and don't rely on a page/component to know it has to call `getUserData` too).
  - note: we use async in the `login` thunk to ensure that we finish getting the user data before letting the rest of the app know that we've finished logging in

## Verification Steps

 

You can run the front end and try logging in to see that redirect works, the user data is fetched successfully, and the error above doesn't show up.

I also modified some old tests to check that `getUserData` is successfully called in the login thunk and created some new integration tests to ensure that the Login page renders correctly based on the `authenticationState`.
